### ISSUE TITLE
Accumulating / Collecting Errors via ErrorListener to handle after the Parse

### ISSUE THREAD

**Issue by jonathandandries** [antlr/antlr4#396]
The ErrorListener mechanism in Antlr4 is great for logging and making decisions about syntax errors as they occur during a parse, but it can get better for batch error handling after the parse is finished. There are a number of reasons you might want to handle errors after the parse finishes, including:
- we need a clean way to programmatically check for errors during a parse and handling them after the fact,
- sometimes one syntax error causes several others (when not recovered in line, for instance), so it can be helpful to group or nest these errors by parent context when displaying output to the user, and you can't know all the errors until the parse is finished,
- you may want to display errors differently to the user depending on how many and how severe they are, for example, a single error that exited a rule or a few errors all recovered in line might just ask the user to fix these local areas - otherwise, you might have the user edit the entire input, and you need to have all the errors to make this determination.

The bottom line is that we can be smarter about reporting and asking users to fix syntax errors if we know the full context in which the errors occurred (including other errors). To do this, I have the following three goals:
1. a full collection of all the errors from a given parse,
2. context information for each error, and
3. severity and recovery information for each error.

I have written code to do goal-1 and goal-2, and I'm looking for help on goal-3. I'm also suggesting some small changes to make goal-1 and goal-2 easier for everyone.

SUMMARY of my suggestions for a future version of ANTLR:
1. Always populate the "RecognitionException e" parameter of ANTLRErrorListener.syntaxError (including the OffendingToken) so that we can collect these exceptions for batch handling after the parse. While your at it, make sure the e.getMessage() is set to return the value currently in the msg parameter.
2. Add a constructor for RecognitionException that includes OffendingToken.
3. Remove the other parameters (aside from "RecognitionException e") in the method signature of ANTLRErrorListener.syntaxError since they will be extraneous and lead to confusion.
4. Add convenience methods in RecognitionException for common stuff such as getCharPositionInLine, getLineNumber, getRuleStack, and the rest of my stuff from my RecognitionExceptionUtil class defined in the StackExchange post linked below. Of course, these will have to check for null and also check that recognizer is of type Parser for some of these methods.
5. When calling ANTLRErrorListener.syntaxError, clone the recognizer so that we don't lose the context when the parse finishes (and we can more easily call getRuleInvocationStack).
6. If you clone the recognizer, you won't need to store the context in RecognitionException (you can get it from the Parser when it's stored as e.getRecognizer). We can then make two changes to e.getCtx(): first, rename it to e.getContext() to make it consistent with Parser.getContext(), and second, make it a convenience method for the recognizer we already have in RecognitionException (checking that recognizer is an instance of Parser).
7. Include information in RecognitionException about the severity of the error and how the parser recovered. This is my goal-3 from above. It would be great to categorize syntax errors by how well the parser handled it. Did this error blow up the entire parse or just show up as a blip in line? How many and which tokens were skipped / inserted?

See my post on StackOverflow for code and more details:

http://stackoverflow.com/questions/20828864/accumulating-collecting-errors-via-errorlistener-to-handle-after-the-parse

**Comment by MarcelSchaible** [comment]
It would be interesting to improve the ANTLR4 error/warning messages like e.g. clang diagnostics:

t.c:7:39: error: invalid operands to binary expression ('int' and 'struct A')
    return y + func(y ? ((SomeA.X + 40) + SomeA) / 42 + SomeA.X : SomeA.X);
                         ~~~~~~~~~~~~~~ ^ ~~~~~

Is someone already working on something similair like this?

**Comment by surister** [comment]
> It would be interesting to improve the ANTLR4 error/warning messages like e.g. clang diagnostics:
> 
> ```
> t.c:7:39: error: invalid operands to binary expression ('int' and 'struct A')
>     return y + func(y ? ((SomeA.X + 40) + SomeA) / 42 + SomeA.X : SomeA.X);
>                          ~~~~~~~~~~~~~~ ^ ~~~~~
> ```
> 
> Is someone already working on something similiar like this?

https://github.com/crate/cratedb-sqlparse/blob/0a122530ca297b118c30ac4e636ec2ab4484127a/cratedb_sqlparse_py/cratedb_sqlparse/parser.py#L47

**Comment by jimidle** [comment]
You' need to do it specifically for your parser, but it is trivial. It won't be built into ANTLR itself as it will always need overriding.

As a guide:

Write your own strategy, inspect the exception stack (different on each target), replace the rule with  a human readable description, remove duplicates, produce a message akin to

While parsing a sub clause in a select statement 'GOUP' was not expected:

SELECT wins FROM LEEDS.GAMES Goup BY Managers....

                             ^^^^

Generate the names by inspecting the exception stack and removing duplicates. Your target language limits your ability to do this of course.

IN short, there is not a universal way to do this, do override the ErrorStrategy and ErrorListener

Jim


a
